javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll

```diff
@@ -0,0 +1,244 @@
+/**
+ * Provides default sources, sinks and sanitizers for reasoning about
+ * unsafe jQuery plugins, as well as extension points for adding your
+ * own.
+ */
+
+import javascript
+private import semmle.javascript.dataflow.InferredTypes
+import semmle.javascript.security.dataflow.Xss
+
+module UnsafeJQueryPlugin {
+  private import DataFlow::FlowLabel
+
+  /**
+   * A data flow source for unsafe jQuery plugins.
+   */
+  abstract class Source extends DataFlow::Node {
+    /**
+     * Gets the plugin that this source is used in.
+     */
+    abstract JQueryPluginMethod getPlugin();
+  }
+
+  /**
+   * A data flow sink for unsafe jQuery plugins.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A sanitizer for unsafe jQuery plugins.
+   */
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * An argument that may act as a HTML fragment rather than a CSS selector, as a sink for remote unsafe jQuery plugins.
+   */
+  class AmbiguousHtmlOrSelectorArgument extends DataFlow::Node {
+    AmbiguousHtmlOrSelectorArgument() {
+      exists(JQuery::MethodCall call |
+        call.interpretsArgumentAsSelector(this) and call.interpretsArgumentAsHtml(this)
+      ) and
+      // the $-function in particular will not construct HTML for non-string values
+      analyze().getAType() = TTString() and
+      // any fixed prefix makes the call unambiguous
+      not exists(DataFlow::Node prefix |
+        DomBasedXss::isPrefixOfJQueryHtmlString(this, prefix) and
+        prefix.mayHaveStringValue(_)
+      )
+    }
+  }
+
+  /**
+   * Holds for jQuery plugin definitions of the form `$.fn.<pluginName> = <plugin>`.
+   */
+  private predicate jQueryPluginDefinition(string pluginName, DataFlow::Node plugin) {
+    exists(DataFlow::PropRead fn, DataFlow::PropWrite write |
+      fn = jquery().getAPropertyRead("fn") and
+      (
+        write = fn.getAPropertyWrite()
+        or
+        exists(ExtendCall extend, DataFlow::SourceNode source |
+          fn.flowsTo(extend.getDestinationOperand()) and
+          source = extend.getASourceOperand() and
+          write = source.getAPropertyWrite()
+        )
+      ) and
+      plugin = write.getRhs() and
+      (
+        pluginName = write.getPropertyName() or
+        write.getPropertyNameExpr().flow().mayHaveStringValue(pluginName)
+      )
+    )
+  }
+
+  /**
+   * Gets an function that is registered as a jQuery plugin method at `def`.
+   */
+  private DataFlow::FunctionNode getAJQueryPluginMethod(
+    DataFlow::TypeBackTracker t, DataFlow::Node def
+  ) {
+    t.start() and
+    jQueryPluginDefinition(_, def) and
+    result.flowsTo(def)
```

Comment:
The return type should generally be `SourceNode` in a recursive type tracking predicate.

With the current return type of `FunctionNode` we won't track any further than `getALocalSource()` since e.g. a `ParameterNode` or `PropRead` can't be included in the base case of the predicate.

A better return type can be set on the "end" predicate, which I'd recommend always defining explicitly.